### PR TITLE
New Update on cdk_framework design against adot test framework

### PR DESCRIPTION
**Pull request by Danielzolty**
**Description:** <Describe what has changed.>

Because I originally made the new PR by making the 2 different clusters stacks and working backwards, I just want to illustrate what the code in the PR has done so far so there is no confusion:
  1. Validated the configuration file
         a. This means that the config file is being checked to make sure it has the necessary fields (no extra fields) and all 
              the field value types are of string (no one is passing in an integer for any of the values).
  2. Create the Interfaces
         a. This is so that after the config file is validated, we cast each cluster in file to ClusterInterface, and if it is 
             launch_type of ec2 , then it will be casted to ec2Interface.
  3. Based on the launch_type, either ec2ClusterStack  or fargateClusterStack gets called and all necessary parameters 
      are passed in where the respective cluster is then deployed.

What is not done yet:
  1. Validating the field values of each cluster being passed in the config file. What I mean by that is for example, making 
     sure the version value is between 1.18 and 1.21.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import { Vpc } from 'aws-cdk-lib/aws-ec2';
+import { KubernetesVersion} from 'aws-cdk-lib/aws-eks';
+import { ManagedPolicy, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
+
+const supportedNodeSizes = new Set(['medium', 'large', 'xlarge', '2xlarge', '4xlarge', '8xlarge', '12xlarge', '16xlarge', '24xlarge', 'metal']);
+const supportedT4gInstances = new Set(['nano', 'micro', 'small', 'medium', 'large', 'xlarge', '2xlarge'])
+
+
+export class EC2Stack extends Stack {
+  cluster : eks.Cluster | eks.FargateCluster
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts]:
```suggestion
  cluster : eks.Cluster
```

File: cdk_framework/EKS/lib/cluster-deployment.ts [hunk 2]
```diff
@@ -21,31 +22,43 @@ export function deployClusters(app: cdk.App) : Map<string, ClusterStack> {
...
+    //       name: String(cluster['name']),
+    //       vpc: vpcStack.vpc,
+    //       version: versionKubernetes,
+    //       env: {
+    //         region: REGION
+    //       },
+    //     })
+    //   }
+    //   eksClusterMap.set(cluster['name'], stack)
+    // }
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/cluster-deployment.ts]: Why is this commented out? Is this dead code? If so, please remove.

File: cdk_framework/EKS/lib/interfaces/ec2cluster-interface [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import {ClusterInterface} from "./cluster-interface"
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/interfaces/ec2cluster-interface]: This file name is missing a type extension

File: cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts (hunk 1 above)
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts]:
```suggestion
  cluster : eks.Cluster
```

File: cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    name: string;
+    vpc: Vpc;
+    version: KubernetesVersion;
+    ec2_instance: string;
+    node_size: string
+}
+
+
+function validateNodeSize(size: string, instance: string){
+    if(!size || size === null || size === 'null' || size === ''){
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts]:
```suggestion
    if(size.length < 1){
```

Comment (bryan-aguilar) [reply cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts]: Would this make more sense?

File: cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        if(!supportedNodeSizes.has(adjustedSize)){
+            throw new Error('Node size is not one of the options listed here https://www.amazonaws.cn/en/ec2/instance-types/')
+        }
+        if(instance === 'm5' && adjustedSize === 'medium'){
+            throw new Error('CPU architecture and node size are not compatible')
+        }
+        if(instance === 'm6g' && adjustedSize === '24xlarge'){
+            throw new Error('CPU architecture and node size are not compatible')
+        }
+    }
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts]: nit: Checking that the node size is in the `supportedNodeSize` set and then using a switch statement may be easier to read here.

File: cdk_framework/EKS/lib/stacks/fargate-cluster-stack.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
+import { Stack, StackProps, aws_eks as eks} from 'aws-cdk-lib';
+import { Construct } from 'constructs';
+import { Vpc } from 'aws-cdk-lib/aws-ec2';
+import { KubernetesVersion} from 'aws-cdk-lib/aws-eks';
+import { Role } from 'aws-cdk-lib/aws-iam';
+
+
+export class FargateStack extends Stack {
+  cluster : eks.Cluster | eks.FargateCluster
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/stacks/fargate-cluster-stack.ts]:
```suggestion
  cluster : eks.FargateCluster
```

File: cdk_framework/EKS/lib/stacks/fargate-cluster-stack.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import { Stack, StackProps, aws_eks as eks} from 'aws-cdk-lib';
+import { Construct } from 'constructs';
+import { Vpc } from 'aws-cdk-lib/aws-ec2';
+import { KubernetesVersion} from 'aws-cdk-lib/aws-eks';
+import { Role } from 'aws-cdk-lib/aws-iam';
+
+
+export class FargateStack extends Stack {
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/stacks/fargate-cluster-stack.ts:8]: Does this stack require any sort of IAM policies like the ec2 worker roles?

File: cdk_framework/EKS/test/validation.test.ts [hunk 8]
```diff
@@ -1,425 +1,425 @@
...
+//                 assert(false)
+//             } catch(error){
+//                 expect(error).toBeInstanceOf(Error);
+//             }
+//         } else {
+//             validateClustersConfig(fields.data)
+//             expect((fields.data)).toEqual(fields.expectedOut);
+//         }
+//     })
+// )
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/test/validation.test.ts]: Why is this file here? Is this dead code? If so, please remove.

File: cdk_framework/EKS/lib/app.ts [hunk 9]
```diff
@@ -1,12 +1,13 @@
...
 import * as cdk from 'aws-cdk-lib';
-import { ClusterStack } from './stacks/cluster-stack';
 import { deployClusters } from './cluster-deployment';
+import { FargateStack } from './stacks/fargate-cluster-stack';
+import { EC2Stack } from './stacks/ec2-cluster-stack';
 
 const app = new cdk.App();
 
-let clusterMap = new Map<string, ClusterStack>()
+let clusterMap = new Map<string, FargateStack | EC2Stack>()
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/app.ts]: `Map<string, FargateStack | EC2Stack>()` gets reused a lot. Can we use a type alias instead of repeating this everywhere? Find a good place to define
```
type clusterMapType = Map<string, FargateStack | EC2Stack>
```

File: cdk_framework/EKS/lib/stacks/fargate-cluster-stack.ts (hunk 7 above)
Comment (Danielzolty) [reply cdk_framework/EKS/lib/stacks/fargate-cluster-stack.ts:8]: There is a default role provided assumed by anyone with permissions in the same account

File: cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts (hunk 5 above)
Comment (Danielzolty) [reply cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts]: I'm going to redo the validation because now I am using interfaces. So going to delete this function.

File: cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts (hunk 4 above)
Comment (Danielzolty) [reply cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts]: I'll look into it. I plan on redoing the validation because I am using interfaces now.

File: cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts (hunk 5 above)
Comment (Danielzolty) [reply cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts]: The issue is that there are 2 sets because the node sizes for t4g is so different from the other ones, I decided to put it a whole different set. Let me know if you still think I should change it from this though.

File: cdk_framework/EKS/package.json [hunk 10]
```diff
@@ -26,13 +26,16 @@
...
-    "@aws-cdk/aws-ec2": "^1.161.0",
-    "@aws-cdk/aws-eks": "^1.161.0",
-    "@aws-cdk/aws-iam": "^1.161.0",
-    "@aws-cdk/core": "^1.161.0",
+    "@aws-cdk/aws-ec2": "1.161.0",
+    "@aws-cdk/aws-eks": "1.161.0",
+    "@aws-cdk/aws-iam": "1.161.0",
+    "@aws-cdk/core": "1.161.0",
+    "ajv": "8.11.0",
+    "ajv-errors": "3.0.0",
```
Comment (rapphil) [inline cdk_framework/EKS/package.json]: Why did you remove the carat from some dependencies?
https://github.com/npm/node-semver#caret-ranges-123-025-004

Please use carat in all dependencies unless you have a good reason for not doing it.

Comment (Danielzolty) [reply cdk_framework/EKS/package.json]: I had an issue with my dependencies not lining up, where I was getting the error illustrated in this stackoverflow post - https://stackoverflow.com/questions/66572044/how-to-resolve-aws-cdk-error-argument-of-type-function-is-not-assignable-to-p#:~:text=37,to%20avoid%20different - so I followed the solution given.

Comment (Danielzolty) [reply cdk_framework/EKS/package.json]: Now everything works, so I'll try re-adding the up-carats

**Review by rapphil — APPROVED** [review]

File: cdk_framework/EKS/lib/cluster-deployment.ts [hunk 11]
```diff
@@ -21,30 +25,49 @@ export function deployClusters(app: cdk.App) : Map<string, ClusterStack> {
...
+      let stack;
+      const clusterInterface = cluster as ClusterInterface
+      if(clusterNameSet.has(clusterInterface.name)){
+        throw new Error(`Cluster name ${clusterInterface.name} is shared by two different clusters`)
+      }
+      clusterNameSet.add(clusterInterface.name)
+      const versionKubernetes = eks.KubernetesVersion.of(clusterInterface.version);
+      if(clusterInterface.launch_type === 'ec2'){
+        const ec2Cluster = cluster as ec2ClusterInterface
+        validateInterface(ec2Cluster)
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/cluster-deployment.ts:49]: Can these `validateInterface` calls be done once before the `if` statement?

File: cdk_framework/EKS/lib/config/cluster-config/clusters.yml [hunk 12]
```diff
@@ -1,16 +1,9 @@
...
-      fargate:
-    version: 1.21
-  t4gCluster01:
-    launch_type: 
-      ec2:
-        ec2_instance: m5
-        node_size: large
-    version: 1.21
+  - name: ec2Cluster
+    version: "1.21"
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/config/cluster-config/clusters.yml:4]: Why does the `version` field require double quotes?

File: cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts [hunk 13]
```diff
@@ -32,40 +33,32 @@ export class ClusterStack extends Stack {
...
-      });
-    } 
+
+    const instance_type = props.instance_type.toLowerCase()
+
+    this.cluster = new eks.Cluster(this, props.name, {
+      clusterName: props.name,
+      vpc: props.vpc,
+      vpcSubnets: [{subnetType: ec2.SubnetType.PUBLIC}],
+      defaultCapacity: 0,  // we want to manage capacity our selves
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts]:
```suggestion
      defaultCapacity: 0,  // we want to manage capacity ourselves
```

File: cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts [hunk 14]
```diff
@@ -32,40 +33,32 @@ export class ClusterStack extends Stack {
...
+      clusterLogging: logging,
     
+    });
+    this.cluster.addNodegroupCapacity(`ng-${instance_type}`, {
+        instanceTypes: [new ec2.InstanceType(instance_type)],
+        minSize: 2,
+        nodeRole: workerRole
+    })
     this.cluster.awsAuth.addMastersRole(Role.fromRoleName(this, 'eks_admin_role', 'Admin'))
+    this.cluster.awsAuth.addMastersRole(Role.fromRoleName(this, 'he', 'Admin'))
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts]: What is the purpose of this master role?

File: cdk_framework/EKS/lib/utils/validate-config-schema.ts [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+                type: "object",
+                properties: {
+                    name: {
+                        type: "string",
+                        errorMessage: {
+                            type: 'Name must be a string'
+                        }
+                    },
+                    version:  {
+                        type: "string",
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/utils/validate-config-schema.ts:22]: Should version be a string or number type?

File: cdk_framework/EKS/lib/utils/validate-interface-schema.ts [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+            throw new Error('CPU architecture and node size are not compatible. Check here for compatibility options: https://www.amazonaws.cn/en/ec2/instance-types/')
+        }
+        if(instance === 'm6g' && adjustedSize === '24xlarge'){
+            throw new Error('CPU architecture and node size are not compatible. Check here for compatibility options: https://www.amazonaws.cn/en/ec2/instance-types/')
+        }
+    }
+
+    return adjustedSize
+
+}
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/lib/utils/validate-interface-schema.ts]: Please add a newline

File: cdk_framework/EKS/lib/config/cluster-config/clusters.yml (hunk 12 above)
Comment (Danielzolty) [reply cdk_framework/EKS/lib/config/cluster-config/clusters.yml:4]: So that it is of type string.  https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-eks-cluster.html#cfn-eks-cluster-version This is a link that Raphael sent me that seems to support this.

File: cdk_framework/EKS/lib/utils/validate-config-schema.ts (hunk 15 above)
Comment (Danielzolty) [reply cdk_framework/EKS/lib/utils/validate-config-schema.ts:22]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-eks-cluster.html#cfn-eks-cluster-version This is a link that Raphael sent me that seems to support this.

File: cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts (hunk 14 above)
Comment (Danielzolty) [reply cdk_framework/EKS/lib/stacks/ec2-cluster-stack.ts]: Oh shoot. That was a mistake, I was playing with things. Good catch!

File: cdk_framework/EKS/lib/cluster-deployment.ts (hunk 11 above)
Comment (Danielzolty) [reply cdk_framework/EKS/lib/cluster-deployment.ts:49]: The reason I do it after is because I want to potentially cast to `ec2ClusterInterface`, so that the validation will also check the `instance_type`

File: cdk_framework/EKS/README.md [hunk 17]
```diff
@@ -25,8 +25,11 @@ The `tsconfig.json` file is used to tell TypeScript how to configure the project
 Since the code base in written in TypeScript, the CDK library has to be downloaded using Node. 
 
 1. Make sure you have Node, so that you can use `npm` control. 
-2. Download from EKS directory the AWS CDK Library by typing `npm install aws-cdk-lib`. 
-3. In order to use the linter, the eslint dependency needs to be downloaded. This could be done by calling `npm install --save-dev eslint @typescript-eslint/parser @typescript-eslint/eslint-plugin`. 
+2. Download from EKS directory the AWS CDK Library by typing `npm install aws-cdk-lib`.
+3. Download `yaml-schema-validator` by typing in command line `npm install yaml-schema-validator`.
+4. Download ajv library by typing in command line `npm install ajv`.
+5. Download ajv errors by typing in command line `npm install ajv-errors`.
```
Comment (bryan-aguilar) [inline cdk_framework/EKS/README.md:31]: doesn't a single `npm install` command do this?

**Review by bryan-aguilar — APPROVED** [review]
